### ISSUE TITLE
cannot multiply `&&EdwardsBasepointTable` by `&Scalar`

### ISSUE THREAD

**Issue by Kofituo** [mcginty/snow#146]
Hi there, I'm facing this issue when compiling my Substrate program:
```
error[E0369]: cannot multiply `&&EdwardsBasepointTable` by `&Scalar`
   --> /Users/kofiotuo/.cargo/registry/src/github.com-1ecc6299db9ec823/snow-0.9.0/src/resolvers/default.rs:132:47
    |
132 |         let point = (&ED25519_BASEPOINT_TABLE * &self.privkey).to_montgomery();
    |                      ------------------------ ^ ------------- &Scalar
    |                      |
    |                      &&EdwardsBasepointTable
    |
help: `*` can be used on `&EdwardsBasepointTable` if you dereference the left-hand side
    |
132 |         let point = (*&ED25519_BASEPOINT_TABLE * &self.privkey).to_montgomery();
    |                      +

For more information about this error, try `rustc --explain E0369`.
error: could not compile `snow` due to previous error

**Comment by evilrobot-01** [comment]
Also just encountered this, seems the latest version of `curve25519-dalek` `v4.0.0-rc.0` released 18 hours ago is causing the issue.

Manually downgrading to the previous version resolved it for me: 
```
cargo update -p curve25519-dalek@4.0.0-rc.0 --precise 4.0.0-pre.5
```

**Comment by tarcieri** [comment]
In the `rc.0` release, you'll probably want to switch to the new `mul_base` method:

```rust
let point = EdwardsPoint::mul_base(&self.privkey).to_montgomery();
```

This abstracts over various fixed-base scalar multiplication implementations based on what optimizations are available, including using the basepoint tables.

**Comment by tarcieri** [comment]
Here's a PR which adds a first-class API for this: https://github.com/dalek-cryptography/curve25519-dalek/pull/503

**Comment by thomaseizinger** [comment]
Likely, the culprit here is that `snow` depends on `curve25519-dalek` with the implicit caret (`^`) operator which allows `cargo` to look for semver-compatible versions: https://github.com/mcginty/snow/blob/dd50e62077407de730421fc66b757ccbd480d87f/Cargo.toml#L48

Pre-releases are [compared lexically](https://semver.org/#spec-item-11), an `rc` happens to have a higher precedence than `pre`.

Would it possible to get a hot fix for `snow` that pins the `curve25519-dalek` version like this:

```toml
curve25519-dalek = { version = "=4.0.0-pre.2", optional = true }
```

**Comment by thomaseizinger** [comment]
Submitted a PR here: https://github.com/mcginty/snow/pull/147

**Comment by mcginty** [comment]
Published a patch fix (`v0.9.1`) on to a new  `0.9` branch https://github.com/mcginty/snow/tree/0.9, as the main branch has API-breaking changes. Should be live on crates now!
